### DISCUSSION TITLE
How to improve reaction drawing?

### DISCUSSION THREAD

**Discussion by paconius** [rdkit/rdkit#6677] [Q&A] [answered]
Hi all. I'm playing around with drawing chemical reactions using RDKit and I would love some feedback on how I can improve the look of the output. Consider the example below. 
![image](https://github.com/rdkit/rdkit/assets/309160/2cade1e1-7d33-474b-aae6-79ba83ba3b74)

 I would like to improve 4 things:
1. In the first structure, the bonds in the nitro group are overlapping the 'N' and one of the 'O' atom labels.
2. In the second structure, there is a bond overlapping the tertiary 'N' atom label
3. In the second structure, the hydrogen on the alcohol oxygen is missing
4. in the third and final structure the single and double bonds on the nitro group are again overlapping the atom labels

Please let me know if you have recommedations to help me here in the context of needing to keep the molecules drawn in the orientation specified in the rxnblock.

Edit: for comparison, using ACS1996 formatting, the ChemDraw rendering looks like this. This would be close to what I am trying to acheive.
![image](https://github.com/rdkit/rdkit/assets/309160/b428b568-35c0-4d74-98ed-faf634cc64b4)

```
from rdkit import Chem
from rdkit.Chem import Draw, AllChem
from rdkit.Chem.Draw import rdMolDraw2D, rdDepictor
import rdkit
from IPython.display import SVG
print(rdkit.__version__) #2023.03.3
rxn = AllChem.ReactionFromRxnBlock("""$RXN

      Mrv1920  083120231704

  2  1
$MOL

  Mrv1920 08312317042D          

 10 10  0  0  0  0            999 V2000
  -15.5505    1.5229    0.0000 F   0  0  0  0  0  0  0  0  0  0  0  0
  -15.0237    0.8880    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  -15.3101    0.1144    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  -14.7833   -0.5206    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  -13.9701   -0.3817    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  -13.6836    0.3920    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  -14.2105    1.0268    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  -13.1108    1.9393    0.0000 O   0  5  0  0  0  1  0  0  0  0  0  0
  -13.9240    1.8006    0.0000 N   0  3  0  0  0  4  0  0  0  0  0  0
  -14.4508    2.4354    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
  2  1  1  0  0  0  0
  3  2  4  0  0  0  0
  2  7  4  0  0  0  0
  4  3  4  0  0  0  0
  5  4  4  0  0  0  0
  6  5  4  0  0  0  0
  7  6  4  0  0  0  0
  9  7  1  0  0  0  0
  9  8  1  0  0  0  0
 10  9  2  0  0  0  0
M  CHG  2   8  -1   9   1
M  END
$MOL

  Mrv1920 08312317042D          

 13 14  0  0  0  0            999 V2000
   -8.2168    0.5454    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
   -8.9969    0.8140    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -8.9825    1.6389    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -8.1935    1.8801    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -7.7203    1.2043    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
   -6.8955    1.1899    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -6.4955    0.4684    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -6.4706    1.8971    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -5.6457    1.8827    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -5.2457    1.1611    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -5.6706    0.4540    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -4.4458   -0.2820    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
   -5.2707   -0.2676    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  2  1  4  0  0  0  0
  1  5  4  0  0  0  0
  3  2  4  0  0  0  0
  4  3  4  0  0  0  0
  5  4  4  0  0  0  0
  6  5  1  0  0  0  0
  7 11  4  0  0  0  0
  9  8  4  0  0  0  0
 10  9  4  0  0  0  0
 11 10  4  0  0  0  0
 13 11  1  0  0  0  0
 13 12  1  0  0  0  0
  8  6  4  0  0  0  0
  6  7  4  0  0  0  0
M  END
$MOL

  Mrv1920 08312317042D          

 22 24  0  0  0  0            999 V2000
    9.5033    1.4166    0.0000 O   0  5  0  0  0  1  0  0  0  0  0  0
    8.1633    1.9132    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    8.6900    1.2781    0.0000 N   0  3  0  0  0  4  0  0  0  0  0  0
    8.4036    0.5053    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    8.9294   -0.1313    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    8.6450   -0.9054    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    7.8317   -1.0420    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    7.3032   -0.4078    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    7.5908    0.3651    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    7.0640    1.0000    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
    6.2508    0.8612    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    5.7240    1.4961    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    6.0104    2.2698    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    5.4836    2.9047    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    4.6704    2.7659    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    4.9107    1.3574    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    4.3839    1.9923    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    3.5707    1.8535    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
    2.9804    2.4298    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.2498    2.0465    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.3886    1.2333    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    3.2049    1.1140    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
  3  1  1  0  0  0  0
  3  2  2  0  0  0  0
  4  3  1  0  0  0  0
  5  4  4  0  0  0  0
  4  9  4  0  0  0  0
  6  5  4  0  0  0  0
  7  6  4  0  0  0  0
  8  7  4  0  0  0  0
  9  8  4  0  0  0  0
 10  9  1  0  0  0  0
 17 16  4  0  0  0  0
 15 17  4  0  0  0  0
 11 12  1  0  0  0  0
 12 13  4  0  0  0  0
 13 14  4  0  0  0  0
 14 15  4  0  0  0  0
 16 12  4  0  0  0  0
 17 18  1  0  0  0  0
 18 19  4  0  0  0  0
 19 20  4  0  0  0  0
 20 21  4  0  0  0  0
 22 18  4  0  0  0  0
 21 22  4  0  0  0  0
 11 10  1  0  0  0  0
M  CHG  2   1  -1   3   1
M  END
""")
reactant1 = rxn.GetReactantTemplate(0)

drawer = rdMolDraw2D.MolDraw2DSVG(-1,-1)
dopts = drawer.drawOptions()
dopts.bondLineWidth = 1.0 # default is 2.0
Draw.SetACS1996Mode(dopts,Draw.MeanBondLength(reactant1))
drawer.DrawReaction(rxn)
drawer.FinishDrawing()

import IPython.display
i = IPython.display.SVG(drawer.GetDrawingText())
display(i)
```

**Comment by paconius** [comment]
Some progress thanks to the tried and true method of chaging random stuff until something works (Monte Carlo?). Changing the bond length supplied to Draw.SetACS1996Mode fixes all of the overlapping bond issues I was observing. A 0.7 multiplier was arbitrarily picked to give a nice looking result. I also bumped up the line thickness using the draw options. The new image looks like this:
![image](https://github.com/rdkit/rdkit/assets/309160/045e5ba6-7510-4e30-afc4-339499ff59f1)

Everything looking great except the missing hydrogens on heteroatoms. Would love to hear ideas about how to accomplish that.

```
drawer = rdMolDraw2D.MolDraw2DSVG(-1,-1)
dopts = drawer.drawOptions()
Draw.SetACS1996Mode(dopts,Draw.MeanBondLength(reactant1)*0.70)
dopts.bondLineWidth = 1.5 # default is 2.0
drawer.DrawReaction(rxn)
drawer.FinishDrawing()
```

**Reply by DavidACosgrove** [reply, marked as answer]
Hi @paconius, I'm just looking at it now.  The reason why scaling the mean bond length works is that the ACS mode works on 14.4 pixels per "Angstrom" where "Angstrom" is one unit of the coordinate frame of the atoms read in.  When it calculates reaction coordinates, RDKit makes a mean bond length of 1.5 "Angstrom" and that looks reasonable on this scale factor.  The scale factor itself is part of the ACS drawing guidelines.  The reaction block you have has a mean bond length of ~0.8, which is why it comes out small. The font size is fixed, so the characters are then drawn over the bonds.  By telling the code that the mean bond length is smaller than it really is, the scale factor (14.4 / meanBondLength) is then increased, and you get the picture you see.

As for drawing the H atoms, that's not really my area of expertise, but I think it might be because the reactants are queries rather than molecules, and for those only explicit H atoms are drawn.  On that basis the ChemDraw drawing would be incorrect, however, so perhaps that explanation is wrong.

**Reply by paconius** [reply]
Thank you David!  Your point about how the mols are "queries rather than molecules" is something I can look into. How do you know that this is the case? Is there a way to check a mol object to see if it is a query vs a molecule?

**Reply by paconius** [reply]
> RDKit makes a mean bond length of 1.5 "Angstrom" and that looks reasonable on this scale factor

Your comment here confused me becuase the coords in those RXN blocks were generated by RDKit (even though the pasted version above looks like Marvin since I was playing around with them in Marvin). How can the mean bond length be 1 for these structures if RDKit defaults to 1.5? I tracked it down to the observation that if you generate coordinates in RDKit after calling
`Chem.rdDepictor.SetPreferCoordGen(True)`
then the bond lengths will be around 1. However if you set this to False, or don't call it at all, the bond lengths will be close to 1.5

**Comment by DavidACosgrove** [comment]
I am extrapolating from SMIRKS where it is the case - the reagents are SMARTS, the products are SMILES.

**Reply by paconius** [reply]
I think you are correct about the problem arising from the mols being querymols instead of regular-mols. I can confirm this by iterating through all atoms in the reaction and printing their python type. I get 'rdkit.Chem.rdchem.QueryAtom' for all atoms! So now I just need to convert all the atoms from rdkit.Chem.rdchem.QueryAtom to rdkit.Chem.rdchem.Atom; however I can't find any info on how to do this. I'll see if this thread gets any more attention and if not I will start a new thread.

**Reply by paconius** [reply]
Hold up, I figured out a hack to get this to work. You can covert a molecule comprised of queryatoms to a molecule comprised of regular atoms by round-tripping it though a molblock. Presumably you could round-trip thorugh some other format, but I need to preserve the atomic coordinates. The code below takes the existing reaction and performs this conversion of each reactant and product and then builds a new reaction from that that can be fed into the draw routine. A bit harebrained, but it seems to work and produces the following output:
![image](https://github.com/rdkit/rdkit/assets/309160/267ca562-ea45-4bac-a2b2-b0a0613d8b21)


```
new_rxn = AllChem.ChemicalReaction()	
for mol in rxn.GetReactants():
	mol = Chem.MolFromMolBlock(Chem.MolToMolBlock(mol))
	new_rxn.AddReactantTemplate(mol)
for mol in rxn.GetProducts():
	mol = Chem.MolFromMolBlock(Chem.MolToMolBlock(mol))
	new_rxn.AddProductTemplate(mol)
```
